python/repositories.bzl

```diff
@@ -1,8 +1,164 @@
+# Copyright 2022 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
 """This file contains macros to be called during WORKSPACE evaluation.
 
 For historic reasons, pip_repositories() is defined in //python:pip.bzl.
 """
 
+load("//python/private:toolchains_repo.bzl", "PLATFORMS", "toolchains_repo")
+load("//python/private:versions.bzl", "MINOR_MAPPING", "RELEASE_DATE", "RELEASE_URL", "TOOL_VERSIONS")
+
 def py_repositories():
     # buildifier: disable=print
     print("py_repositories is a no-op and is deprecated. You can remove this from your WORKSPACE file")
+
+########
+# Remaining content of the file is only used to support toolchains.
+########
+_DOC = "Fetch external tools needed for python toolchain."
+_ATTRS = {
+    "platform": attr.string(mandatory = True, values = PLATFORMS.keys()),
+    "python_version": attr.string(mandatory = True, values = TOOL_VERSIONS.keys() + MINOR_MAPPING.keys()),
+}
+
+def _python_repo_impl(rctx):
+    version = rctx.attr.python_version
+    if version in MINOR_MAPPING:
+        version = MINOR_MAPPING[version]
+    release_filename = "cpython-{version}-{platform}-pgo+lto-{release_date}.tar.zst".format(
+        release_date = RELEASE_DATE,
+        version = version,
+        platform = rctx.attr.platform,
+    )
+    url = "{release_url}/{release_filename}".format(
+        release_url = RELEASE_URL,
+        release_filename = release_filename,
+    )
+    rctx.download(
+        url = url,
+        integrity = TOOL_VERSIONS[version][rctx.attr.platform],
```

Comment:
If sha is an empty string, bazel will emit a warning, which the user can chose to act upon. This way we can avoid hardcoding shas of the releases and we can avoid coupling the releases of `indygreg` Python interpreter and `rules_python`.